File: sudoku.c

#include <stdio.h>
#include <string.h>
#include <ctype.h>
#include <time.h>

#define len(a) (sizeof(a)/sizeof*(a))

int find(
		int (**rs)[9],
		int *n,
		int (*data)[9],
		int (*row)[10],
		int (*col)[10],
		int *(*blo)[9],
		int x,
		int y);
int display(int (**rs)[9],int n);

int main(int argc,char *argv[])
{
	unsigned long start=clock();

	/* main data */
	static int data[9][9];
	static int row[9][10],col[9][10],blockset[9][10];
	static int *blo[9][9];
	static int (*rs[10])[9],n;

	char c;
	int i,j,k,l;

	/* input */
	for(i=0;i<9*9;)
		if(isdigit(c=getchar()))
			data[0][i]=c-'0',i++;

	/* init */
	for(i=0;i<9;i++)
		for(j=0;j<9;j++)
			blo[i][j]=blockset[i/3*3+j/3];
	for(i=0;i<9;i++)
		for(j=0;j<9;j++){
			k=data[i][j];
			row[i][k]=col[j][k]=blo[i][j][k]=1;
		}

	/* search */
	find(rs,&n,data,row,col,blo,0,0);
	if(n)
		display(rs,n);
	start=clock()-start;
	printf("time used: %lu ms.\n",start);

	while(n--)
		free(rs[n]);
	//free(*rs);
	return 0;
}

int find(
		int (**rs)[9],
		int *n,
		int (*data)[9],
		int (*row)[10],
		int (*col)[10],
		int *(*blo)[9],
		int x,
		int y){
	/* is it the end */
	if(x==9){
		printf("find it!\n");
		rs[*n]=(int(*)[9])malloc(sizeof(int)*9*9);
		memcpy(rs[*n],data,sizeof(int)*9*9);
		(*n)++;
		return 1;
	}
	/* get the coordinate of the next point */
	int ntx,nty;
	if(y==8)
		ntx=x+1,nty=0;
	else
		ntx=x,nty=y+1;
	/* the process of recursion */
	if(data[x][y])
		return find(rs,n,data,row,col,blo,ntx,nty);
	else{
		int i,tmp;
		for(i=1;i<10;i++)
			if(row[x][i]==0
				&&col[y][i]==0
				&&blo[x][y][i]==0){
				data[x][y]=i;
				row[x][i]=col[y][i]=blo[x][y][i]=1;
				tmp=find(rs,n,data,row,col,blo,ntx,nty);
				data[x][y]=row[x][i]=col[y][i]=blo[x][y][i]=0;
				if(tmp)
					return 1;
			}
		return 0;
	}
}

int display(int (**rs)[9],int n){
	int i,j,k;
	for(i=0;i<n;i++){
		for(j=0;j<9;j++){
			for(k=0;k<9;k++)
				printf("%2d",rs[i][j][k]);
			putchar('\n');
		}
		putchar('\n');
	}

	return 0;
}

File: sudoku.py

class sudoku(object):
	#fil=lambda data: filter(lambda x : x!=0, data)
	def __init__(self,data):
		self.bo=[ data[i*9:i*9+9] for i in range(9) ]
		self.r=[ self.fil(i) for i in self.bo]
		temp=[ data[i::9] for i in range(9) ]
		self.c=[ self.fil(i) for i in temp ]

		b=[[],[],[],]
		for i in range(0,9,3):
			for j in range(0,9,3):
				temp=[]
				for ii in range(3):
					for jj in range(3):
						temp.append(self.bo[i+ii][j+jj])
				b[i//3].append(self.fil(temp))
		self.b=[]
		for i in range(9):
			self.b.append([])
		for i in range(9):
			for j in range(9):
				self.b[i].append(b[i//3][j//3])
		self.rs=[]
	def put(self,x,y,f):
		if (f not in self.r[x] 
				and f not in self.c[y] 
				and f not in self.b[x][y] ):
			self.bo[x][y]=f
			self.r[x].add(f)
			self.c[y].add(f)
			self.b[x][y].add(f)
			return True
		return False
	def remove(self,x,y,f):
		self.bo[x][y]=0
		self.r[x].remove(f)
		self.c[y].remove(f)
		self.b[x][y].remove(f)
	def display(self):
		for i in self.bo:
			print i
		if self.rs:
			print('the result is')
			for i in self.rs:
				print i
	def fil(self,data):
		rs=set(data)
		try:
			rs.remove(0)
		except:
			pass
		finally:
			return rs

def find(a,x,y):
	if x==9:
		print('find it')
		a.rs=copy.deepcopy(a.bo)
		return True
		pass
	
	xx,yy=(x,y+1) if y<9-1 else (x+1,0)
	
	if a.bo[x][y]!=0:
		return find(a,xx,yy)
		pass
	if x<9:
		for i in range(1,10):
			if(a.put(x,y,i)):
				rs=find(a,xx,yy)
				a.remove(x,y,i)
				if rs:
					return True
		return False


import copy
import time

bo=[]
while len(bo)<9*9 :
	temp=list(raw_input())
	temp=filter(lambda x: x.isdigit(), temp)
	bo+=temp
del bo[9*9:]
bo=map(int,bo)

a=sudoku(bo)
start=time.clock()
find(a,0,0)
start=time.clock()-start
a.display()
rs='time used:%.4f s'%(start)
print rs

